File: js/traversals.js
const inOrderTraversal = (tree) => {
  if (tree == null) {
    return;
  }

  inOrderTraversal(tree.left);
  console.log(tree.value);
  inOrderTraversal(tree.right);
};

const preOrderTraversal = (tree) => {
  if (tree == null) {
    return;
  }

  console.log(tree.value);
  preOrderTraversal(tree.left);
  preOrderTraversal(tree.right);
};

const postOrderTraversal = (tree) => {
  if (tree == null) {
    return;
  }

  postOrderTraversal(tree.left);
  postOrderTraversal(tree.right);
  console.log(tree.value);
};

module.exports = { inOrderTraversal, preOrderTraversal, postOrderTraversal };

File: c/traversals.c
#include <stdlib.h>
#include "traversals.h"

void in_order_traversal(Node_ptr head, Displayer displayer)
{
  if (head == NULL)
  {
    return;
  }

  in_order_traversal(head->left, displayer);
  (*displayer)(head->value);
  in_order_traversal(head->right, displayer);
}

void pre_order_traversal(Node_ptr head, Displayer displayer)
{
  if (head == NULL)
  {
    return;
  }

  (*displayer)(head->value);
  pre_order_traversal(head->left, displayer);
  pre_order_traversal(head->right, displayer);
}

void post_order_traversal(Node_ptr head, Displayer displayer)
{
  if (head == NULL)
  {
    return;
  }

  post_order_traversal(head->left, displayer);
  post_order_traversal(head->right, displayer);
  (*displayer)(head->value);
}
